browser/src/ws.ts

```diff
@@ -0,0 +1,107 @@
+/* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+import { ConnectionConfig } from "./mqtt";
+import WebsocketStream = require("websocket-stream");
+import * as Crypto from "crypto-js";
+
+export interface WebsocketOptions {
+    headers?: { [index: string]: string };
+    protocol?: string;
+}
+
+export interface AWSCredentials {
+    aws_region?: string,
+    aws_access_id: string,
+    aws_secret_key: string,
+    aws_sts_token?: string
+}
+
+function zero_pad(n: number) {
+    return (n > 9) ? n : '0' + n.toString();
+}
+
+function canonical_time() {
+    const now = new Date();
+    return `${now.getUTCFullYear()}${zero_pad(now.getUTCMonth() + 1)}${zero_pad(now.getUTCDate())}T` +
+        `${zero_pad(now.getUTCHours())}${zero_pad(now.getUTCMinutes())}${zero_pad(now.getUTCSeconds())}Z`;
+}
+
+function canonical_day(time : string = canonical_time()) {
+    return time.substring(0, time.indexOf('T'));
+}
+
+function make_signing_key(credentials: AWSCredentials, day: string, service_name: string) {
+    const hash_opts = { asBytes: true };
+    let hash = Crypto.HmacSHA256(day, 'AWS4' + credentials.aws_secret_key, hash_opts);
+    hash = Crypto.HmacSHA256(credentials.aws_region || '', hash, hash_opts);
+    hash = Crypto.HmacSHA256(service_name, hash, hash_opts);
+    hash = Crypto.HmacSHA256('aws4_request', hash, hash_opts);
+    return hash;
+}
+
+function sign_url(method: string,
+    url: URL,
+    credentials: AWSCredentials,
+    service_name: string,
+    time: string = canonical_time(),
+    day: string = canonical_day(time),
+    payload: string = '') {
+    
+    const signed_headers = 'host';
+    const canonical_headers = `host:${url.hostname.toLowerCase()}\n`;
+    const payload_hash = Crypto.SHA256(payload, { asBytes: true });
+    const canonical_params = url.search.replace(new RegExp('^\\?'), '');
```

Comment:
What is the transformation being done here?

Reply:
```url.search``` always contains the leading ```?``` from the query params. I suppose this could be written
```js
if (url.search[0] === '?') {
    url.search = url.search.substring(1);
}
```

Reply:
ah makes sense now